File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/event/EventRegistry.java
package top.likoslupus.cellulosesz.api.event;

import top.likoslupus.cellulosesz.api.service.Registration;

import java.util.function.Consumer;

public interface EventRegistry {

    default <T> Registration listen(
            Class<T> eventType,
            Consumer<T> listener
    ) {
        return listen(eventType, listener, "global");
    }

    <T> Registration listen(
            Class<T> eventType,
            Consumer<T> listener,
            String owner
    );

    default <T extends CancellableEvent> boolean fireCancellable(T event) {
        fire(event);
        return !event.cancelled();
    }

    <T> void fire(T event);

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/event/EventRegistry.kt
package top.likoslupus.cellulosesz.api.event

import top.likoslupus.cellulosesz.api.service.Registration
import java.util.function.Consumer

public interface EventRegistry {

    public fun <T : Any> listen(
        eventType: Class<T>,
        listener: Consumer<T>
    ): Registration =
        listen(eventType, listener, "global")

    public fun <T : Any> listen(
        eventType: Class<T>,
        listener: Consumer<T>,
        owner: String
    ): Registration

    public fun <T : CancellableEvent> fireCancellable(event: T): Boolean {
        fire(event)
        return !event.cancelled()
    }

    public fun <T : Any> fire(event: T)

}
